homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,193 @@
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_CUSTOM,
+    CONF_FACTOR,
+    CONF_GROUP,
+    CONF_KEY,
+    CONF_UNIT,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    PLATFORMS,
+    PYSMA_COORDINATOR,
+    PYSMA_OBJECT,
+    PYSMA_REMOVE_LISTENER,
+    PYSMA_SENSORS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _parse_legacy_options(entry: ConfigEntry, sensor_def: pysma.Sensors) -> None:
+    """Parse legacy configuration options.
+
+    This will parse the legacy CONF_SENSORS and CONF_CUSTOM configuration options
+    to support deprecated yaml config from platform setup.
+    """
+
+    # Add sensors from the custom config
+    sensor_def.add(
+        [
+            pysma.Sensor(o[CONF_KEY], n, o[CONF_UNIT], o[CONF_FACTOR], o.get(CONF_PATH))
+            for n, o in entry.data.get(CONF_CUSTOM).items()
+        ]
+    )
+
+    # Parsing of sensors configuration
+    config_sensors = entry.data.get(CONF_SENSORS)
+    if config_sensors:
+        # Find and replace sensors removed from pysma
+        # This only alters the config, the actual sensor migration takes place in _migrate_old_unique_ids
+        for sensor in config_sensors.copy():
+            if sensor in pysma.LEGACY_MAP:
+                config_sensors.remove(sensor)
+                config_sensors.append(pysma.LEGACY_MAP[sensor]["new_sensor"])
+
+        # Only sensors from config should be enabled
+        for sensor in sensor_def:
+            sensor.enabled = sensor.name in config_sensors
+
+
+async def _migrate_old_unique_ids(
+    hass: HomeAssistant, entry: ConfigEntry, sensor_def: pysma.Sensors
+) -> None:
+    """Migrate legacy sensor entity_id format to new format."""
+    entity_registry = er.async_get(hass)
+
+    # Create list of all possible sensor names
+    possible_sensors = list(
+        set(
+            entry.data.get(CONF_SENSORS)
+            + [s.name for s in sensor_def]
+            + list(pysma.LEGACY_MAP)
+        )
+    )
+
+    for sensor in possible_sensors:
+        if sensor in sensor_def:
+            pysma_sensor = sensor_def[sensor]
+            original_key = pysma_sensor.key
+        elif sensor in pysma.LEGACY_MAP:
+            # If sensor was removed from pysma we will remap it to the new sensor
+            legacy_sensor = pysma.LEGACY_MAP[sensor]
+            pysma_sensor = sensor_def[legacy_sensor["new_sensor"]]
+            original_key = legacy_sensor["old_key"]
+        else:
+            _LOGGER.error("%s does not exist", sensor)
+            continue
+
+        # Find entity_id using previous format of unique ID
+        entity_id = entity_registry.async_get_entity_id(
+            "sensor", "sma", f"sma-{original_key}-{sensor}"
+        )
+
+        if entity_id:
```

Comment:
```suggestion
        if not entity_id:
            continue
```

Then outdent below.


This one is small though so it probably doesn't matter so much

Reply:
Fixed:
https://github.com/home-assistant/core/commit/2ad7ad8c9aae7286e00dd578cbf2342eac6bcbc4#diff-a7bdb8a0a15589eb3d6f0a9b41a9131ef4a2563166a67a1536b973a7534adcfeL105-R112